### PR TITLE
Port ItemsRepeater changes from WinUI

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Updates `ItemsRepeater` and associated classes with changes made in https://github.com/microsoft/microsoft-ui-xaml up to commit `007ab33a66642acb3be343f5aaf35df650513df5`

### CODE REVIEW

File: src/Avalonia.Layout/NonVirtualizingLayoutContext.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿// This source file is adapted from the WinUI project.
+// (https://github.com/microsoft/microsoft-ui-xaml)
+//
+// Licensed to The Avalonia Project under MIT License, courtesy of The .NET Foundation.
+
+namespace Avalonia.Layout
+{
+    public abstract class NonVirtualizingLayoutContext : LayoutContext
```
Comment (MarchingCube) [inline src/Avalonia.Layout/NonVirtualizingLayoutContext.cs]: Documentation comment missing.

**Comment by MarchingCube** [comment]
Gave your PR a try. Looks like that performance of `UniformGrid` layout is much better now, despite no commit mentioning any perf changes.

Stack layout is still completely broken when using randomized heights, do you know if this is an issue with UWP implementation or it is on our side?

**Comment by grokys** [comment]
> Stack layout is still completely broken when using randomized heights, do you know if this is an issue with UWP implementation or it is on our side?

Is the problem with the scrollbar jumping around as you scroll? For that I think we need to implement [scroll anchoring](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.iscrollanchorprovider). I was planning to do that as part of the work on using `ItemsRepeater` as our `ItemsPresenter.

**Comment by jmacato** [comment]
@grokys yes, i think it's the same issue that i raised when we were testing random height items in itemsrepeater. You told me that we need scroll anchoring for it

**Comment by MarchingCube** [comment]
@grokys Yeah, scrollbar jumps around like crazy and sometimes ends up being stuck jumping around and showing no items.

File: src/Avalonia.Layout/NonVirtualizingLayoutContext.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Layout/NonVirtualizingLayoutContext.cs]: Yeah, this is just a placeholder class for now, I'm working on adding the stuff necessary for non-virtualizing layouts in another PR. I can add a doc comment though.

**Comment by grokys** [comment]
Working on a scroll anchoring implementation now. It will be a separate PR.

**Review by MarchingCube — APPROVED** [review]
LGTM.

**Pull request merged**